experiments/targeting.go

```diff
@@ -0,0 +1,357 @@
+//
+//     {
+//        EQ: {
+//             field: <field_name>
+//             value: <accepted_value>
+//         }
+//     }
+//
+// The expected input to this constructor from the above example would be::
+//
+//     {
+//         field: <field_name>,
+//         value: <accepted_value>
+//     }
+//
+type EqualNode struct {
+	fieldName      string
+	acceptedValues []interface{}
+}
+
+// NewEqualNode parses the underlying input into an EqualNode.
+func NewEqualNode(inputNodes map[string]interface{}) (Targeting, error) {
+	if len(inputNodes) != 2 {
+		return nil, TargetingNodeError("EqualNode expects exactly two fields")
+	}
+	acceptedKey, ok := inputNodes["field"]
+	if !ok {
+		return nil, TargetingNodeError("EqualNode expects input key 'field'")
+	}
+
+	value, valueOK := inputNodes["value"]
+	values, valuesOK := inputNodes["values"]
+	if !valueOK && !valuesOK {
+		return nil, TargetingNodeError("EqualNode expects input key 'value' or 'values'.")
+	}
+
+	var acceptedValues []interface{}
+	if valuesOK {
+		acceptedValues = values.([]interface{})
+	} else if valueOK {
+		acceptedValues = []interface{}{value}
+	}
+	return &EqualNode{
+		fieldName:      strings.ToLower(acceptedKey.(string)),
+		acceptedValues: acceptedValues,
+	}, nil
+}
+
+// Evaluate returns true if the given attribute has the expected value.
+func (n *EqualNode) Evaluate(inputs map[string]interface{}) bool {
+	candidateValue := inputs[n.fieldName]
+	switch n := candidateValue.(type) {
+	case int, float64:
+		candidateValue = json.Number(fmt.Sprintf("%v", n))
+	}
+	for _, value := range n.acceptedValues {
+		if candidateValue == value {
+			return true
+		}
+	}
+	return false
+}
+
+// NotNode is a boolean 'not' operator and negates the child node.
+type NotNode struct {
+	child Targeting
+}
+
+// NewNotNode parses the underlying input into an NotNode.
+func NewNotNode(inputNodes map[string]interface{}) (*NotNode, error) {
+	if len(inputNodes) != 1 {
+		return nil, TargetingNodeError("NotNode expects exactly one field")
+	}
+	node, err := parseNode(inputNodes)
+	if err != nil {
+		return nil, err
+	}
+	return &NotNode{
+		child: node,
+	}, nil
+}
+
+// Evaluate returns the negation of the child's evaluation.
+func (n *NotNode) Evaluate(inputs map[string]interface{}) bool {
+	return !n.child.Evaluate(inputs)
+}
+
+// OverrideNode is an override to the targeting and can always return true or
+// false.
+type OverrideNode struct {
+	ReturnValue bool
+}
+
+// NewOverrideNode parses the underlying input into an OverrideNode.
+func NewOverrideNode(inputNode interface{}) *OverrideNode {
+	returnValue, ok := inputNode.(bool)
+	if !ok {
+		returnValue = false
+	}
+	return &OverrideNode{
+		ReturnValue: returnValue,
+	}
+}
+
+// Evaluate returns the configured boolean return value.
+func (n *OverrideNode) Evaluate(inputs map[string]interface{}) bool {
+	return n.ReturnValue
+}
+
+// ComparisonNode is a non-equality comparison operators (gt, ge, lt, le).
```

Comment:
🔕 Does it make more sense to merge `EqualNode` into `ComparisonNode` as well? `ne` is already handled here.

Reply:
I think the `EqualNode` allows for a wider variety of comparisons, string, integer, bool fields etc. where as comparison seems to be reserved for numbers, but I actually need to verify this. If that's not the case then `ComparisonNode` needs to become more flexible, too.